bigquery_magics/graph_server.py

```diff
@@ -0,0 +1,220 @@
+    # does not even get called unless spanner_graphs has already been confirmed
+    # to exist upstream.
+    from google.cloud.spanner_v1.types import StructType, Type, TypeCode
+    from spanner_graphs.conversion import (
+        columns_to_native_numpy,
+        prepare_data_for_graphing,
+    )
+
+    try:
+        column_name = None
+        column_value = None
+        for key, value in query_results.items():
+            if column_name is None:
+                if not isinstance(key, str):
+                    raise ValueError(f"Expected outer key to be str, got {type(key)}")
+                if not isinstance(value, dict):
+                    raise ValueError(
+                        f"Expected outer value to be dict, got {type(value)}"
+                    )
+                column_name = key
+                column_value = value
+            else:
+                # TODO: Implement multi-column support.
+                raise ValueError(
+                    "Query has multiple columns - graph visualization not supported"
+                )
+        if column_name is None or column_value is None:
+            raise ValueError(
+                "query result with no columns is not supported for graph visualization"
+            )
+
+        fields: List[StructType.Field] = [
+            StructType.Field(name=column_name, type=Type(code=TypeCode.JSON))
+        ]
+        data = {column_name: []}
+        rows = []
+        for value_key, value_value in column_value.items():
+            if not isinstance(value_key, str):
+                raise ValueError(f"Expected inner key to be str, got {type(value_key)}")
+            if not isinstance(value_value, str):
+                raise ValueError(
+                    f"Expected inner value to be str, got {type(value_value)}"
+                )
+            row_index = int(value_key)
+            row_json = json.loads(value_value)
+
+            if row_index != len(data[column_name]):
+                raise ValueError(
+                    f"Unexpected row index; expected {len(data[column_name])}, got {row_index}"
+                )
+            data[column_name].append(row_json)
+            rows.append([row_json])
+
+        d, ignored_columns = columns_to_native_numpy(data, fields)
+
+        graph: DiGraph = prepare_data_for_graphing(incoming=d, schema_json=None)
+
+        nodes = []
+        for node_id, node in graph.nodes(data=True):
+            nodes.append(node)
+
+        edges = []
+        for from_id, to_id, edge in graph.edges(data=True):
+            edges.append(edge)
+
+        return {
+            "response": {
+                "nodes": nodes,
+                "edges": edges,
+                "schema": None,
+                "rows": rows,
+                "query_result": data,
+            }
+        }
+    except Exception as e:
+        return {"error": getattr(e, "message", str(e))}
+
+
+class GraphServer:
```

Comment:
It's a bit strange to see so many static methods. Doesn't seem particularly Pythonic.

Perhaps some docstrings explaining this is a kind of singleton could help? But even then, this doesn't really follow a typical singleton pattern. A module with some global state would be a bit more Pythonic in terms of doing this sort of thing (https://stackoverflow.com/a/31887/101923).

One pattern that might be a bit easier to test is actually one that we use in this file, where we don't attempt to make a true singleton, but instead have a particular global object we use: https://github.com/googleapis/python-bigquery-magics/blob/50cc9060a353c495ba547efa2568e63c962d1505/bigquery_magics/config.py#L206